crates/volta-core/src/tool/package/fetch.rs

```diff
@@ -80,9 +77,95 @@ fn load_cached_distro(file: &Path, shasum_file: &Path) -> Option<Box<dyn Archive
     Tarball::load(distro).ok()
 }
 
-fn fetch_remote_distro(spec: tool::Spec, url: &str, path: &Path) -> Fallible<Box<dyn Archive>> {
-    debug!("Downloading {} from {}, to {}", &spec, &url, path.display());
-    Tarball::fetch(url, path).with_context(download_tool_error(spec, url.to_string()))
+fn fetch_remote_distro(
+    path: &Path,
+    name: &str,
+    details: &PackageDetails,
+    session: &mut Session,
+) -> Fallible<Box<dyn Archive>> {
+    ensure_containing_dir_exists(&path).with_context(|_| ErrorDetails::ContainingDirError {
+        path: path.to_path_buf(),
+    })?;
+
+    let dir = path.parent().unwrap();
+
+    let command = npm_pack_command_for(name, &details.version.to_string()[..], session, dir)?;
+    debug!("Running command: `{:?}`", command);
+
+    debug!(
+        "Downloading {} via npm pack to {}",
+        tool_version(name, details.version.to_string()),
+        dir.to_str().unwrap()
+    );
+    let spinner = progress_spinner(&format!(
+        "Downloading {}",
+        tool_version(name, details.version.to_string()),
+    ));
+    let output = command.output()?;
+    spinner.finish_and_clear();
+
+    if !output.status.success() {
+        debug!(
+            "Command failed, stderr is:\n{}",
+            String::from_utf8_lossy(&output.stderr).to_string()
+        );
+        debug!("Exit code is {:?}", output.status.code());
+        throw!(ErrorDetails::NpmPackFetchError {
+            package: tool_version(name, details.version.to_string()),
+        });
+    }
+
+    let filename = String::from_utf8_lossy(&output.stdout);
+    // The output from `npm pack` contains a newline, so we'll trim it here.
+    let trimmed_filename = filename.trim();
+
+    if trimmed_filename.is_empty() {
+        throw!(ErrorDetails::NpmPackUnpackError {
+            package: tool_version(name, details.version.to_string())
+        });
+    }
+
+    let tarball_from_npm_pack = dir.join(trimmed_filename.to_string());
+
+    if !tarball_from_npm_pack.exists() {
+        throw!(ErrorDetails::NpmPackUnpackError {
+            package: tool_version(name, details.version.to_string())
+        });
+    }
+
+    debug!(
+        "Moving the tarball from {:?} to the expected path {:?}",
+        tarball_from_npm_pack, path
+    );
+    rename(tarball_from_npm_pack, path).with_context(|_| ErrorDetails::NpmPackUnpackError {
+        package: tool_version(name, details.version.to_string()),
+    })?;
```

Comment:
We probably won't need to do this most of the time, since generally the file name we get from `npm pack` should match the one we generate internally. Given that, can we do a check before the `rename` to make sure that the two file paths are different?

`if tarball_from_npm_pack != path {` should work, it looks like `Path` implements `PartialEq`

Reply:
Ah, that makes sense. My use case is a scoped package, where the naming is different. Will do! 👍